File: content/snippets/scala2/interview_lcci/01/02_check_permutation.scala
object Lcci0102CheckPermutation {
  def checkPermutation(a: String, b: String): Boolean = {
    if (a.length != b.length) {
      return false
    }

    val freq = Array.fill(256)(0)
    a.foreach(ch => freq(ch) += 1)

    b.foreach { ch =>
      freq(ch) -= 1
      if (freq(ch) < 0) {
        return false
      }
    }

    true
  }

  def main(args: Array[String]): Unit = {
    println(s"abcde vs edcba -> ${checkPermutation("abcde", "edcba")}")
    println(s"abc vs abz -> ${checkPermutation("abc", "abz")}")
  }
}

File: content/snippets/go/interview_lcci/01/02_check_permutation.go
package main

import "fmt"

func checkPermutation(a string, b string) bool {
	if len(a) != len(b) {
		return false
	}

	freq := [256]int{}
	for i := 0; i < len(a); i++ {
		freq[a[i]]++
	}
	for i := 0; i < len(b); i++ {
		freq[b[i]]--
		if freq[b[i]] < 0 {
			return false
		}
	}
	return true
}

func main() {
	fmt.Printf("abcde vs edcba -> %t\n", checkPermutation("abcde", "edcba"))
	fmt.Printf("abc vs abz -> %t\n", checkPermutation("abc", "abz"))
}
